packages/protocol/scripts/foundry/make-release.ts

```diff
@@ -0,0 +1,805 @@
+  console.log('loadContractArtifact', contractName, artifactPath)
+  const artifact = readJsonSync(artifactPath) as ForgeArtifact
+  const sourceFiles = Object.keys(artifact.metadata.sources)
+  return {
+    contractName,
+    abi: artifact.abi as Abi,
+    bytecode: artifact.bytecode.object,
+    address: '0x0' as ViemAddress,
+    sourceFiles,
+  }
+}
+
+const findContractArtifacts = (baseDir: string, contractArtifactPathsMap: Map<string, string>) => {
+  const entries = readdirSync(baseDir, { withFileTypes: true })
+  for (const entry of entries) {
+    if (!entry.isDirectory() || !entry.name.endsWith('.sol')) {
+      continue
+    }
+    const contractSolDirPath = join(baseDir, entry.name as string)
+    const filesInSolDir = readdirSync(contractSolDirPath, { withFileTypes: true })
+
+    for (const fileEntry of filesInSolDir) {
+      if (!fileEntry.isFile() || !fileEntry.name.endsWith('.json')) {
+        continue
+      }
+
+      const contractName = basename(fileEntry.name as string, '.json')
+      const artifactFilePath = join(contractSolDirPath, fileEntry.name as string)
+
+      try {
+        const content = readJsonSync(artifactFilePath)
+        if (content.abi && content.bytecode) {
+          contractArtifactPathsMap.set(contractName, artifactFilePath)
+        } else {
+          console.warn(`Skipping non-ABI/bytecode file: ${artifactFilePath}`)
+        }
+      } catch (e) {
+        console.warn(`Skipping non-JSON or unreadable file: ${artifactFilePath}`)
+      }
+    }
+  }
+}
+
+const linkLibraries = (
+  contractViemArtifact: ViemContract,
+  contractDependencies: string[],
+  addresses: ContractAddresses
+): void => {
+  if (!contractViemArtifact.bytecode.includes('__')) {
+    return
+  }
+
+  if (contractDependencies.length === 0) {
+    console.error(
+      `No dependencies found for ${contractViemArtifact.contractName}. Skipping library linking.`
+    )
+    return
+  }
+
+  for (const dep of contractDependencies) {
+    if (addresses.addresses.has(dep)) {
+      const libAddress = addresses.get(dep).replace('0x', '')
+      const libSourceFilePath = contractViemArtifact.sourceFiles.find((file) =>
+        file.includes(`${dep}.sol`)
+      )
+
+      if (!libSourceFilePath) {
+        console.error(
+          `Could not determine sourceFilePath for library ${dep}. Skipping Foundry placeholder replacement for it in ${contractViemArtifact.contractName}.`
+        )
+        return
+      }
+
+      const stringToHash = `${libSourceFilePath}:${dep}`
+      const hashed = keccak256(toHex(new TextEncoder().encode(stringToHash)))
+      const placeholderHash = hashed.substring(2, 2 + 34)
+
+      const placeholderRegexDollar = new RegExp(`__\\$${placeholderHash}\\$__`, 'g')
+      if (contractViemArtifact.bytecode!.match(placeholderRegexDollar)) {
+        contractViemArtifact.bytecode = contractViemArtifact.bytecode!.replace(
+          placeholderRegexDollar,
+          libAddress
+        ) as Hex
```

Comment:
`replace()` will replace single occurrence of `__dep__` as library in bytecode.
What would happen if there are couple references to given library in the bytecode?
```suggestion
        contractViemArtifact.bytecode = contractViemArtifact.bytecode!.replaceAll(
          placeholderRegexDollar,
          libAddress
        ) as Hex
```